File: python/46_permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # Variables
        results = []
        permutation = []

        # Check Input
        if not nums:
            return results

        # Execute function
        self.find_permutations(results, permutation, nums)

        return results


    # Define function
    def find_permutations(self, results, permutation, nums):
        if not nums:
            results.append(permutation[:])
            return

        for i in range(len(nums)):
            permutation.append(nums[i])
            self.find_permutations(results, permutation, nums[:i] + nums[(i + 1):])
            permutation.pop()

File: go/46_permutations.go
func permute(nums []int) [][]int {
	var results [][]int
	var permutation []int

	find_permutations(nums, permutation, &results)
	return results
}

func find_permutations(nums []int, permutation []int, results *[][]int) {
	if len(nums) == 0 {
		*results = append(*results, permutation)
		return
	}

	for i, n := range nums {
		find_permutations(
			append(append([]int{}, nums[:i]...), nums[(i+1):]...),
			append(permutation, n),
			results,
		)
	}
}